honeycombio/data_source_columns.go

```diff
@@ -0,0 +1,90 @@
+package honeycombio
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	honeycombio "github.com/honeycombio/terraform-provider-honeycombio/client"
+	"github.com/honeycombio/terraform-provider-honeycombio/honeycombio/internal/hashcode"
+)
+
+func dataSourceHoneycombioColumns() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceHoneycombioColumnsRead,
+
+		Schema: map[string]*schema.Schema{
+			"dataset": {
+				Type:     schema.TypeString,
+				Optional: false,
+				Required: true,
+			},
+			"starts_with": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"names": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"types": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func dataSourceHoneycombioColumnsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*honeycombio.Client)
+
+	dataset := d.Get("dataset").(string)
+
+	columns, err := client.Columns.List(ctx, dataset)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	value, ok := d.GetOk("starts_with")
+	if ok {
+		startsWith := value.(string)
+
+		for i := len(columns) - 1; i >= 0; i-- {
+			if !strings.HasPrefix(columns[i].KeyName, startsWith) {
+				columns = append(columns[:i], columns[i+1:]...)
+			}
+		}
+	}
+
+	names := make([]string, len(columns))
+	types := make([]string, len(columns))
+
+	for i, d := range columns {
+		names[i] = d.KeyName
+
+		switch *d.Type {
+		case honeycombio.ColumnTypeString:
+			types[i] = "string"
+		case honeycombio.ColumnTypeFloat:
+			types[i] = "float"
+		case honeycombio.ColumnTypeInteger:
+			types[i] = "integer"
+		case honeycombio.ColumnTypeBoolean:
+			types[i] = "boolean"
+		}
```

Comment:
I think we should drop the `types` list as mentioned above, but you can avoid the switch here altogether as the underlying type is a string anyway :)

```suggestion
        types[i] = *d.Type.(string)
```